File: erlang/binary-trees/binary_trees.erl
-module(binary_tree).
-export([main/0, make_tree/1, check_tree/1]).

-record(tree, {left, right}).

make_tree(0) ->
    null;
make_tree(Depth) ->
    DepthMinusOne = Depth - 1,
    #tree{left = make_tree(DepthMinusOne),
          right = make_tree(DepthMinusOne)}.

check_tree(null) ->
    0;
check_tree(#tree{left = Left, right = Right}) ->
    1 + check_tree(Left) + check_tree(Right).

main() ->
    N = 21,
    Depth = 22,
    lists:foreach(
        fun(_) ->
            Tree = make_tree(Depth),
            TotalNodes = check_tree(Tree),
            io:format("Tree of depth ~p has ~p nodes.~n", [Depth, TotalNodes])
        end,
        lists:seq(1, N)).

File: python/binary-trees/binary-trees.py
import random
import sys
#import multiprocessing as mp

def make_tree(depth):
    """
    Recursively generates a binary tree of specified depth.
    Each node is a tuple (left_child, right_child).
    A leaf node is represented as (None, None).
    """
    if depth > 0:
        depth -= 1
        return (make_tree(depth), make_tree(depth))
    return (None, None)

def check_tree(node):
    """
    Recursively calculates and returns the total number of nodes in the tree,
    including the current node, by traversing the tree.
    """
    (left, right) = node
    if left is None:  # Leaf node
        return 1
    else:
        return 1 + check_tree(left) + check_tree(right)

def main():
    """
    Main function to drive the tree creation and checking for 21 trees.
    """
    for _ in range(45):
        depth = 45
        tree = make_tree(depth)
        total_nodes = check_tree(tree)
        print(f'Tree of depth {depth} has {total_nodes} nodes.')

if __name__ == '__main__':
    main()